pennylane/qnodes/base.py

```diff
@@ -311,6 +357,113 @@ def _append_op(self, op):
+                qml.Rot(w[0], w[1], w[2], wires=[1])
+                qml.CRX(-a, wires=[0, 1])
+
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+            qfunc(3.4, [1.2, 3.4, 5.6])
+
+        In this example, ``_make_variables`` will return the following :class:`~.variable.Variable` instances
+
+        .. code-block:: python
+            >>> qfunc._make_variables([3.4, [1.2, 3.4, 5.6]], {})
+            ["a", ["w[0]", "w[1]", "w[2]"]], {}
+
+        where the Variable instances are replaced with their name for readability.
+
+        Args:
+            args (tuple[Any]): Positional arguments passed to the quantum function.
+                During the construction we are not concerned with the numerical values, but with
+                the nesting structure.
+                Each positional argument is replaced with a :class:`~.variable.Variable` instance.
+            kwargs (dict[str, Any]): Auxiliary arguments passed to the quantum function.
+        """
+        # Get the name of the qfunc's arguments
+        full_argspec = inspect.getfullargspec(self.func)
+
+        # args
+        variable_name_strings = []
+        for variable_name, variable_value in zip(full_argspec.args, args):
+            variable_name_strings.append(self._determine_structured_variable_name(variable_value, variable_name))
+
+        # varargs
+        len_diff = len(args) - len(full_argspec.args)
+        if len_diff > 0:
+            for idx, variable_value in enumerate(args[-len_diff:]):
+                variable_name = "{}[{}]".format(full_argspec.varargs, idx)
+
+                variable_name_strings.append(self._determine_structured_variable_name(variable_value, variable_name))
+
+        arg_vars = [Variable(idx, name) for idx, name in enumerate(_flatten(variable_name_strings))]
+        self.num_variables = len(arg_vars)
+
+        # arrange the newly created Variables in the nested structure of args
+        arg_vars = unflatten(arg_vars, args)
+
+        # kwargs
+        variable_name_strings = {}
+        kwarg_vars = {}
+        for variable_name in full_argspec.kwonlyargs:
+            if variable_name in kwargs:
+                variable_value = kwargs[variable_name]
+            else:
+                variable_value = full_argspec.kwonlydefaults[variable_name]
+
+            if isinstance(variable_value, np.ndarray):
+                variable_name_string = np.empty_like(variable_value, dtype=object)
+
+                for index in np.ndindex(*variable_name_string.shape):
+                    variable_name_string[index] = "{}[{}]".format(variable_name, ",".join([str(i) for i in index]))
+
+                kwarg_variable = [Variable(idx, name=name, is_kwarg=True) for idx, name in enumerate(_flatten(variable_name_string))]
+            else:
+                kwarg_variable = Variable(0, name=variable_name, is_kwarg=True)
+
+            kwarg_vars[variable_name] = kwarg_variable
+
+        return arg_vars, kwarg_vars
```

Comment:
This makes it much easier to read and parse! Also, it's great that a developer-facing method has such a detailed docstring, it will make it much easier for other developers.

My only concern is the additional use of `_flatten()` above every time construct is called. `_flatten` can be _very_ slow for large arrays, so this could create a bottleneck/slowdown

Reply:
I implemented Variable caching so that `make_variables` is only called once